romancal/source_detection/tests/test_source_detection_step.py

```diff
@@ -0,0 +1,252 @@
+        wfi_image = testutil.mk_level2_image(shape=shape)
+        wfi_image.data = u.Quantity(
+            np.ones(shape, dtype=np.float32), ru.electron / u.s, dtype=np.float32
+        )
+        wfi_image.meta.filename = "filename"
+
+        # add noise to data
+        if noise is not None:
+            wfi_image.data += u.Quantity(
+                noise * np.random.random(shape), ru.electron / u.s, dtype=np.float32
+            )
+
+        # add dq array
+
+        wfi_image.dq = np.zeros(shape, dtype=np.uint32)
+        # construct ImageModel
+        mod = ImageModel(wfi_image)
+
+        return mod
+
+    return _setup
+
+
+@pytest.mark.skipif(
+    os.environ.get("CI") == "true",
+    reason="Roman CRDS servers are not currently available outside the internal "
+    "network",
+)
+def add_random_gauss(arr, x_positions, y_positions, min_amp=200, max_amp=500):
+
+    """Add random 2D Gaussians to `arr` at specified positions,
+    with random amplitudes from `min_amp` to  `max_amp`. Assumes
+    units of e-/s."""
+
+    for i, x in enumerate(x_positions):
+        y = y_positions[i]
+        gauss = Gaussian2DKernel(2, x_size=20, y_size=20).array
+        amp = np.random.randint(200, 700)
+        arr[y - 10 : y + 10, x - 10 : x + 10] += (
+            u.Quantity(gauss, ru.electron / u.s, dtype=np.float32) * amp
+        )
+
+
+@pytest.mark.skipif(
+    os.environ.get("CI") == "true",
+    reason="Roman CRDS servers are not currently available outside the internal "
+    "network",
+)
+def test_source_detection_defaults(setup_inputs):
+
+    """Test SourceDetectionStep with its default parameters. The detection
+    threshold will be chosen based on the image's background level."""
+
+    model = setup_inputs()
+
+    # add in 12 sources, roughly evenly distributed
+    # sort by true_x so they can be matched up to output
+
+    true_x = np.array([20, 25, 30, 35, 40, 45, 50, 55, 60, 70, 80, 82])
+    true_y = np.array([26, 80, 44, 19, 66, 39, 29, 72, 54, 29, 80, 62])
+
+    # at each position, place a 2d gaussian
+    # randomly vary flux from 100 to 500 for each source
+    add_random_gauss(model.data, true_x, true_y)
+
+    # call SourceDetectionStep with default parameters
+    res = SourceDetectionStep.call(model)
+
+    # unpack output catalog array
+    _, xcentroid, ycentroid, flux = res.meta.source_detection.tweakreg_catalog
+
+    # sort based on x coordinate, like input
+    ycentroid = [x for y, x in sorted(zip(xcentroid, ycentroid))]
+    flux = [x for y, x in sorted(zip(xcentroid, flux))]
+    xcentroid = sorted(xcentroid)
+
+    # check that the number of input and output sources are the same
+    assert len(flux) == len(true_x)
+
+    # check that their locations agree
+    # the differences between real/actual are always around 0.5. im assuming
+    # this is because photutils uses zero-indexed pixels so the centers are
+    # (0.5, 0.5)? atol=0.2 seems to be the lowest safe value for this right now.
+
```

Comment:
We do need to get this convention correct.  The important thing is that we choose the same thing as the distortion model.  The gWCS convention is that (0, 0) is the center of the first pixel.  My guess is that the issue is that you're putting these stars in off center in add_random_guess, and that if you change the kernel size to 21x21 (i.e., something odd), this will go away.  (This has bitten me before!)

FWIW, for noise-free symmetric Gaussians, we might guess that agreement could be very good.  Your images have noise and 0.2 pix might be fine.

Reply:
when i changed the kernel size to odd this was resolved, i was also able to reduce the tolerance to 0.1 pixels